modules/OFStateManager/module/src/gentable_handlers.c

```diff
@@ -896,43 +1034,174 @@ hash_key(of_list_bsn_tlv_t *key)
+                                           OF_BSN_GENTABLE_ERROR_UNKNOWN,
+                                           err_txt);
+    } else if (rv < 0) {
+        AIM_LOG_ERROR("%s gentable delete failed: %s",
+                      gentable->name, indigo_strerror(rv));
+        indigo_cxn_send_bsn_gentable_error(cxn_id, obj, 
+                                           gentable->table_id,
+                                           OF_BSN_GENTABLE_ERROR_UNKNOWN,
+                                           err_txt);
+    }
+}
+
+/**
+ * This API is used by iteration and gentable unregister cases.
+ * Both cases don't have the obj
+ * - iteration case: not used for now. Just for passing the build.
+ * - gentable unregister case: Don't care async operation. Just clear the entry.
+ */
 static indigo_error_t
-delete_entry(indigo_cxn_id_t cxn_id, indigo_core_gentable_t *gentable,
-             struct ind_core_gentable_entry *entry, of_desc_str_t err_txt)
+delete_entry(indigo_cxn_id_t cxn_id,
+             indigo_core_gentable_t *gentable,
+             struct ind_core_gentable_entry *entry,
+             of_desc_str_t err_txt)
 {
     indigo_error_t rv;
-    struct ind_core_gentable_checksum_bucket *checksum_bucket;
 
     if (entry->refcount > 0) {
         return INDIGO_ERROR_EXISTS;
     }
 
-    if (gentable->ops->del3 != NULL) {
+    if (gentable->ops->del4 != NULL) {
+        indigo_core_op_context_t op_ctx;
+        op_ctx.cxn_id = cxn_id;
+        op_ctx.obj = NULL;
+        rv = gentable->ops->del4(&op_ctx, gentable->priv, entry->priv, entry->key, err_txt, true);
+    } else if (gentable->ops->del3 != NULL) {
         rv = gentable->ops->del3(cxn_id, gentable->priv, entry->priv, entry->key, err_txt);
     } else if (gentable->ops->del2 != NULL) {
         rv = gentable->ops->del2(cxn_id, gentable->priv, entry->priv, entry->key);
     } else {
         rv = gentable->ops->del(gentable->priv, entry->priv, entry->key);
     }
-    if (rv < 0) {
-        return rv;
-    }
 
-    bighash_remove(gentable->key_hashtable, &entry->key_hash_entry);
-    list_remove(&entry->checksum_links);
+    /* Dont care about the return value. Force to remove it. */
+    delete_entry_gt_update(gentable, entry);
 
-    checksum_bucket = find_checksum_bucket(gentable, &entry->checksum);
+    return rv;
+}
 
-    subtract_checksum(&checksum_bucket->checksum, &entry->checksum);
-    subtract_checksum(&gentable->checksum, &entry->checksum);
+void
+indigo_core_gentable_entry_resume(
+    indigo_core_op_context_t *op_ctx,
+    void *priv,
+    of_desc_str_t err_txt,
+    indigo_error_t rv)
+{
+    indigo_cxn_id_t cxn_id;
+    of_object_t *obj;
 
-    of_object_delete(entry->key);
-    of_object_delete(entry->value);
-    aim_free(entry);
+    if (op_ctx == NULL) {
+        /* incorrect resume call */
+        AIM_LOG_ERROR("%s should have valid op_ctx", __FUNCTION__);
+        return;
+    }
 
-    gentable->num_entries--;
+    aim_free(op_ctx);
```

Comment:
op_ctx is freed before it is used in the following lines.  please double check the unit test; it should not have passed